### PR TITLE
Reference implementation of scopes

### PR DESCRIPTION
**Pull request by guybedford**
This provides a simple implementation of scoped resolution with the various fallback tests.

The gist of the approach is:
* Sort scopes, and scope imports just like imports
* Generalize the matcher into a `resolveImportsMatch` function which is used on scopes
* The first best matching scope is checked for import resolution, falling back to the global imports.

Because of the way scopes are normalized during parsing, we end up needing to check for `parentURL[normalizedScopeURL.length] === '/' || parntURL[normalizedScopeURL.length - 1] === '/'` because scopes may or may not end in '/'. One optimization here might be to normalize scopes to always be with or without the trailing slash (possibly without?), but that seemed to affect quite a few of the parsing tests when I tried it. I could certainly implement that as well though.

Any other thoughts re edge cases very welcome, but I think most are already covered by the base import maps themselves.

### CODE REVIEW

File: reference-implementation/__tests__/resolving-scopes.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+          "lodash-dot/": "./node_modules_2/lodash-es/",
+          "lodash-dotdot": "../node_modules_2/lodash-es/lodash.js",
+          "lodash-dotdot/": "../node_modules_2/lodash-es/"
+        }
+      }
+    }`);
+
+    const subScriptURL = new URL('https://example.com/app.mjs');
+
+    it('should resolve scoped and not cascade', () => {
```
Comment (domenic) [inline reference-implementation/__tests__/resolving-scopes.js]: What does "and not cascade" mean?

File: reference-implementation/__tests__/resolving-scopes.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        "/js": {
+          "lodash-dot": "./node_modules_2/lodash-es/lodash.js",
+          "lodash-dot/": "./node_modules_2/lodash-es/",
+          "lodash-dotdot": "../node_modules_2/lodash-es/lodash.js",
+          "lodash-dotdot/": "../node_modules_2/lodash-es/"
+        }
+      }
+    }`);
+
+    const subScriptURL = new URL('https://example.com/app.mjs');
```
Comment (domenic) [inline reference-implementation/__tests__/resolving-scopes.js]: Maybe `topLevelScriptURL` or `onlyInOneScopeURL`?

File: reference-implementation/__tests__/resolving-scopes.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    it('should resolve scoped and not cascade', () => {
+      expect(resolveUnderTest('lodash-dot')).toMatchURL('https://example.com/app/node_modules_2/lodash-es/lodash.js');
+      expect(resolveUnderTest('lodash-dotdot')).toMatchURL('https://example.com/node_modules_2/lodash-es/lodash.js');
+      expect(resolveUnderTest('lodash-dot/foo')).toMatchURL('https://example.com/app/node_modules_2/lodash-es/foo');
+      expect(resolveUnderTest('lodash-dotdot/foo')).toMatchURL('https://example.com/node_modules_2/lodash-es/foo');
+    });
```
Comment (domenic) [inline reference-implementation/__tests__/resolving-scopes.js]: What about `moment` with the default `scriptURL`? Does it use the `/` scope definition or not? I would think it should, but we need to test, either way.

File: reference-implementation/__tests__/resolving-scopes.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    it('should apply best scope match', () => {
+      expect(resolveUnderTest('moment', subScriptURL)).toMatchURL('https://example.com/node_modules_3/moment/src/moment.js');
+    });
+
+    it('should fallback to imports', () => {
+      expect(resolveUnderTest('moment/foo', subScriptURL)).toMatchURL('https://example.com/node_modules/moment/src/foo');
```
Comment (domenic) [inline reference-implementation/__tests__/resolving-scopes.js]: Good test.

File: reference-implementation/__tests__/resolving-scopes.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      expect(resolveUnderTest('lodash-dot', subScriptURL)).toMatchURL('https://example.com/app/node_modules/lodash-es/lodash.js');
+      expect(resolveUnderTest('lodash-dotdot', subScriptURL)).toMatchURL('https://example.com/node_modules/lodash-es/lodash.js');
+      expect(resolveUnderTest('lodash-dot/foo', subScriptURL)).toMatchURL('https://example.com/app/node_modules/lodash-es/foo');
+      expect(resolveUnderTest('lodash-dotdot/foo', subScriptURL)).toMatchURL('https://example.com/node_modules/lodash-es/foo');
+    });
+
+    it('should still fail for package modules that are not declared', () => {
```
Comment (domenic) [inline reference-implementation/__tests__/resolving-scopes.js]:
```suggestion
    it('should still fail for package-like specifiers that are not declared', () => {
```

File: reference-implementation/lib/parser.js [hunk 6]
```diff
@@ -16,15 +16,9 @@ exports.parseFromString = (input, baseURL) => {
...
+  let sortedAndNormalizedImports = {};
   if ('imports' in parsed) {
-    normalizedImports = normalizeSpecifierMap(parsed.imports, baseURL);
-  }
-
-  const sortedAndNormalizedImports = {};
-  const sortedImportsKeys = Object.keys(normalizedImports).sort(longerLengthThenCodeUnitOrder);
-  for (const key of sortedImportsKeys) {
-    sortedAndNormalizedImports[key] = normalizedImports[key];
+    sortedAndNormalizedImports = normalizeSpecifierMap(parsed.imports, baseURL);
```
Comment (domenic) [inline reference-implementation/lib/parser.js]: Maybe rename to sortAndNormalizeSpecifierMap

**Comment by domenic** [comment]
I'm going to add to this PR to implement no-trailing-slash = exact matching, trailing-slash = prefix matching. For now that seems best to me. But let's keep #73 open and discuss more fully there; we can change after landing this.

File: reference-implementation/__tests__/resolving-scopes.js (hunk 1 above)
Comment (guybedford) [reply reference-implementation/__tests__/resolving-scopes.js]: I see you worked this out! Seems good to have it too.
